File: python/python-leetcode/longest_substring_without_repeating_characters.py
# https://leetcode.com/problems/longest-substring-without-repeating-characters/
# https://leetcode.com/submissions/detail/814252380/
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        char_found = {}
        max_length = 0
        window_start = 0
        for element in range(len(s)):
            if s[element] not in char_found:
                char_found[s[element]] = element
            else:
                while s[element] in char_found:
                    del char_found[s[window_start]]
                    window_start += 1
                char_found[s[element]] = element
            max_length = max(max_length, element - window_start + 1)
        return max_length


sol = Solution()
print(sol.lengthOfLongestSubstring(s="roaafkte"))

File: go/go-leetcode/examples/longest_substring_without_repeating_characters.go
package main

import (
	"fmt"
)

//https://leetcode.com/problems/concatenation-of-array/
//https://leetcode.com/submissions/detail/814913579/

func main() {
	fmt.Println(lengthOfLongestSubstring("abcabcbb"))
}

func lengthOfLongestSubstring(s string) int {
	mapped, maxLength, left := make(map[rune]int), 0, 0
	for i, current := range s {
		if val, okay := mapped[current]; okay == true && val >= left {
			if i-left > maxLength {
				maxLength = i - left
			}
			left = val + 1
		}
		mapped[current] = i
	}
	if len(s)-left > maxLength {
		maxLength = len(s) - left
	}
	return maxLength
}
